src/Gui/DlgCreateNewPreferencePackImp.cpp

```diff
@@ -0,0 +1,156 @@
+        switch (t.type) {
+        case Gui::PreferencePack::Type::Appearance: itemColumns.push_back(tr("Appearance")); break;
+        case Gui::PreferencePack::Type::Behavior: itemColumns.push_back(tr("Behavior")); break;
+        case Gui::PreferencePack::Type::Combination: itemColumns.push_back(tr("Combination")); break;
+        }
+        auto newItem = new QTreeWidgetItem(group, itemColumns);
+        newItem->setCheckState(0, Qt::Checked);
+        if (group->checkState(0) != newItem->checkState(0))
+            group->setCheckState(0, Qt::PartiallyChecked);
+        newItem->setData(0, TemplateRole, QVariant::fromValue(t));
+        group->addChild(newItem);
+    }
+}
+
+std::vector<Gui::PreferencePackManager::TemplateFile> DlgCreateNewPreferencePackImp::selectedTemplates() const
+{
+    std::vector<Gui::PreferencePackManager::TemplateFile> results;
+
+    for (const auto& group : _groups)
+        for (int childIndex = 0; childIndex < group.second->childCount(); ++childIndex)
+            if (auto child = group.second->child(childIndex); child->checkState(0) == Qt::Checked)
+                if (child->data(0, TemplateRole).canConvert<Gui::PreferencePackManager::TemplateFile>())
+                    results.push_back(child->data(0, TemplateRole).value<Gui::PreferencePackManager::TemplateFile>());
+
+    return results;
+}
+
+std::string DlgCreateNewPreferencePackImp::preferencePackName() const
+{
+    return ui->lineEdit->text().toStdString();
+}
+
+void DlgCreateNewPreferencePackImp::onItemChanged(QTreeWidgetItem* item, int column)
+{
+    Q_UNUSED(column);
+    const QSignalBlocker blocker(ui->treeWidget);
+    if (auto group = item->parent(); group) {
+        // Child clicked
+        bool firstItemChecked = false;
+        for (int childIndex = 0; childIndex < group->childCount(); ++childIndex) {
+            auto child = group->child(childIndex);
+            if (childIndex == 0) {
+                firstItemChecked = child->checkState(0) == Qt::Checked;
+            }
+            else {
+                bool thisItemChecked = child->checkState(0) == Qt::Checked;
+                if (firstItemChecked != thisItemChecked) {
+                    group->setCheckState(0, Qt::PartiallyChecked);
+                    return;
+                }
+            }
+        }
+        group->setCheckState(0, firstItemChecked ? Qt::Checked : Qt::Unchecked);
+    }
+    else {
+        // Group clicked:
+        auto groupCheckState = item->checkState(0);
+        for (int childIndex = 0; childIndex < item->childCount(); ++childIndex) {
+            auto child = item->child(childIndex);
+            child->setCheckState(0, groupCheckState);
+        }
+    }
+}
+
+void DlgCreateNewPreferencePackImp::on_lineEdit_textEdited(const QString& text)
+{
+    if (!text.isEmpty())
+        ui->buttonBox->button(QDialogButtonBox::Ok)->setEnabled(true);
+    else
+        ui->buttonBox->button(QDialogButtonBox::Ok)->setEnabled(false);
```

Comment:
Not sure it's general, but to me is more readable 'ui->buttonBox->button(QDialogButtonBox::Ok)->setEnabled(!text.isEmpty())'

Reply:
Good idea, will change.